pkg/apis/traceflow/v1alpha1/types.go

```diff
@@ -0,0 +1,195 @@
+	// Namespace is the destination namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the destination pod, exclusive with destination service.
+	Pod string `json:"pod,omitempty"`
+	// Service is the destination service, exlusive with destination pod.
+	Service string `json:"service,omitempty"`
+	// IP is the destination IP.
+	IP string `json:"IP,omitempty"`
+}
+
+// IPHeader describes spec of an IPv4 header. IPv6 not supported yet.
+type IPHeader struct {
+	// SrcIP is the source IP.
+	SrcIP string `json:"srcIP,omitempty"`
+	// Protocol is the IP protocol.
+	Protocol int32 `json:"protocol,omitempty"`
+	// TTL is the IP TTL.
+	TTL int32 `json:"ttl,omitempty"`
+	// Flags is the flags for IP.
+	Flags int32 `json:"flags,omitempty"`
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	ICMP *ICMPEchoRequestHeader `json:"icmp,omitempty"`
+	UDP  *UDPHeader             `json:"udp,omitempty"`
+	TCP  *TCPHeader             `json:"tcp,omitempty"`
+}
+
+// ICMPEchoRequestHeader describes spec of an ICMP echo request header.
+type ICMPEchoRequestHeader struct {
+	// ID is the ICMPEchoRequestHeader ID.
+	ID int32 `json:"id,omitempty"`
+	// Sequence is the ICMPEchoRequestHeader sequence.
+	Sequence int32 `json:"sequence,omitempty"`
+}
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
+}
+
+// TCPHeader describes spec of a TCP header.
+type TCPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
+	// Flags are flags in the header.
+	Flags int32 `json:"flags,omitempty"`
+}
+
+// Packet includes payload info and headers info.
+type Packet struct {
+	// PayloadString is the content in payload.
+	PayloadString string `json:"payloadString,omitempty"`
+	// PayloadSize is the size of payload.
+	PayloadSize int32 `json:"payloadSize,omitempty"`
+
+	IPHeader        IPHeader        `json:"ipHeader,omitempty"`
+	TransportHeader TransportHeader `json:"transportHeader,omitempty"`
+}
+
+// Status describes current status of the traceflow.
+type Status struct {
+	// Phase is the Traceflow phase.
+	Phase Phase `json:"phase,omitempty"`
+	// DataplaneTag is an ID encoded into packet traceflow header to identify inter-node packet.
+	DataplaneTag uint8 `json:"dataplaneTag,omitempty"`
+	// Results is the collection of all observations on different nodes.
+	Results []NodeResult `json:"results,omitempty"`
+}
+
+type NodeResult struct {
+	// Node is the node of the observation.
+	Node string `json:"node,omitempty"`
+	// Role of the node like sender, receiver, etc.
+	Role string `json:"role,omitempty"`
+	// Timestamp is the timestamp of the observations on the node.
+	Timestamp int64 `json:"timestamp,omitempty"`
```

Comment:
Do all observation on a Node share the same timestamp?

Reply:
+1

Reply:
Yes.

Reply:
Yes, the timestamp is generated when agent generating observation, thus it will be same for one  Node.